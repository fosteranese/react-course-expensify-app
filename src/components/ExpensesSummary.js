import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import getVisibleExpenses from '../selectors/expenses';
import getExpenseTotal from '../selectors/expense-total';

export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
    const sumLiteral = numeral(expenseTotal / 100).format('$0,0.00');
    const expenseLiteral = expenseCount > 1 ? 'expenses' : 'expense';
    
    return (
        <div>
            Viewing {expenseCount} {expenseLiteral} totalling {sumLiteral}
        </div>
    );
};

const mapStateToProps = ({expenses, filters}) => {
    const visibleExpenses = getVisibleExpenses(expenses, filters);

    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: getExpenseTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);