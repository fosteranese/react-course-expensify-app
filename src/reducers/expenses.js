const expenseReducerDefaultState = [];

export default (state = expenseReducerDefaultState, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'EDIT_EXPENSE':
            return state.map((expense) => (expense.id === action.id ? {...expense, ...action.updates} : expense));
        case 'REMOVE_EXPENSE':
            return state.filter(({id})=> (action.id !== id));
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};