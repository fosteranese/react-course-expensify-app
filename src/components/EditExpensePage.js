import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {

  editExpense = (updates) => {
    this.props.editExpense(this.props.expense.id, updates);
    this.props.history.push("/");
  };

  removeExpense = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.editExpense} />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  ),
});

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, updates) => dispatch(editExpense(id, updates)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({id})),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
