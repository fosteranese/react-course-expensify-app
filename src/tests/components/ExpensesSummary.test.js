import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import { filters, altFilters } from '../fixtures/filters';

let wrapper;
beforeEach(() => {
    wrapper = shallow(<ExpensesSummary expenses={expenses} filters={filters} />);
});

test(
    'should render expense summary correclty',
    () => {
        expect(wrapper).toMatchSnapshot();
    }
);

test(
    'should return the correct filter message with more than one expenses',
    () => {
        expect(wrapper.text()).toBe("Viewing 3 expenses totalling $1,141.95");
    }
);

test(
    'should return the correct filter message with 1 expense',
    () => {
        wrapper.setProps({
            filters: {
                ...filters,
                text: 'Rent'
            }
        });
        expect(wrapper.text()).toBe("Viewing 1 expense totalling $1,095.00");
    }
);

test(
    'should return the correct filter message with 0 expense',
    () => {
        wrapper.setProps({
            filters: {
                ...filters,
                text: 'nothing'
            }
        });
        expect(wrapper.text()).toBe("Viewing 0 expense totalling $0.00");
    }
);