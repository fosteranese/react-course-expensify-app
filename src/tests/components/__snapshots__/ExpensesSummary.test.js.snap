// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary correclty 1`] = `
<div>
  Viewing 
  10
   
  expenses
   totalling 
  $15.00
</div>
`;
